### PR TITLE
Chore/bump nextjs and more

### PR DESCRIPTION
**Pull request by Sepehr-Sobhani**
[ZH CARD 1222](https://app.zenhub.com/workspaces/climate-action-secretariat-60ca4121764d710011481ca2/issues/gh/bcgov/cas-cif/1222)

### Major Changes:
- When using events/triggers in unit tests, we need to use `act`like:
```act(() => userEvent.click(screen.getByText(/Create a new Project/i)));```
- we don’t need to use anchor tags inside `<Link />` component.
- Welcome to cypress 12
- Always run e2e tests on the electron browser rather than chrome (cypress uses electron on the CI)

### Notes:
- I had to use `cy.wait` commands to make cypress happy so I'm going to make a bug card to remove those commands (there are lots of spec work commits related to this)
- It is worth spending some time looking at major changes in [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html) and [NextJS 13](https://nextjs.org/docs/upgrading)
- I've removed `teimp` Growthbook feature flag as part of this issue

### CODE REVIEW

**Review by tmastrom — COMMENTED** [review]
There is a warning in the console related to the link component and growthbook when rendering the page initially. Is growthbook feature flagging working with the updated packages? 


`Warning: validateDOMNesting(...): <a> cannot appear as a descendant of <a>.
    at a
    at O (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:31:19811)
    at Link (webpack-internal:///./node_modules/@button-inc/bcgov-theme/esm/Link.js:37:23)
    at a
    at LinkComponent (webpack-internal:///./node_modules/next/dist/client/link.js:143:19)
    at p
    at div
    at div
    at main
    at div
    at div
    at DefaultLayout (webpack-internal:///./components/Layout/DefaultLayout.tsx:45:23)
    at Index (webpack-internal:///./pages/index.tsx:56:29)
    at WiredComponent (webpack-internal:///./node_modules/relay-nextjs/wired/component.js:61:33)
    at RelayEnvironmentProvider (webpack-internal:///./node_modules/react-relay/lib/relay-hooks/RelayEnvironmentProvider.js:20:24)
    at ErrorBoundary (webpack-internal:///./node_modules/@sentry/react/esm/errorboundary.js:41:47)
    at GrowthBookProvider (webpack-internal:///./node_modules/@growthbook/growthbook-react/dist/esm/index.js:170:3)
    at MyApp (webpack-internal:///./pages/_app.tsx:77:24)
    at PathnameContextProviderAdapter (webpack-internal:///./node_modules/next/dist/shared/lib/router/adapters.js:101:25)
    at ErrorBoundary (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:301:63)
    at ReactDevOverlay (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:850:919)
    at Container (webpack-internal:///./node_modules/next/dist/client/index.js:122:5)
    at AppContainer (webpack-internal:///./node_modules/next/dist/client/index.js:304:24)
    at Root (webpack-internal:///./node_modules/next/dist/client/index.js:522:25)`

**Comment by Sepehr-Sobhani** [comment]
> There is a warning in the console related to the link component and growthbook when rendering the page initially. Is growthbook feature flagging working with the updated packages?
> 
> `Warning: validateDOMNesting(...): <a> cannot appear as a descendant of <a>. at a at O (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:31:19811) at Link (webpack-internal:///./node_modules/@button-inc/bcgov-theme/esm/Link.js:37:23) at a at LinkComponent (webpack-internal:///./node_modules/next/dist/client/link.js:143:19) at p at div at div at main at div at div at DefaultLayout (webpack-internal:///./components/Layout/DefaultLayout.tsx:45:23) at Index (webpack-internal:///./pages/index.tsx:56:29) at WiredComponent (webpack-internal:///./node_modules/relay-nextjs/wired/component.js:61:33) at RelayEnvironmentProvider (webpack-internal:///./node_modules/react-relay/lib/relay-hooks/RelayEnvironmentProvider.js:20:24) at ErrorBoundary (webpack-internal:///./node_modules/@sentry/react/esm/errorboundary.js:41:47) at GrowthBookProvider (webpack-internal:///./node_modules/@growthbook/growthbook-react/dist/esm/index.js:170:3) at MyApp (webpack-internal:///./pages/_app.tsx:77:24) at PathnameContextProviderAdapter (webpack-internal:///./node_modules/next/dist/shared/lib/router/adapters.js:101:25) at ErrorBoundary (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:301:63) at ReactDevOverlay (webpack-internal:///./node_modules/next/dist/compiled/@next/react-dev-overlay/dist/client.js:850:919) at Container (webpack-internal:///./node_modules/next/dist/client/index.js:122:5) at AppContainer (webpack-internal:///./node_modules/next/dist/client/index.js:304:24) at Root (webpack-internal:///./node_modules/next/dist/client/index.js:522:25)`

@tmastrom Yes, it should work. do you get this error on the dashboard page?!

File: app/components/Form/ProjectAnnualReportForm.tsx [hunk 1]
```diff
@@ -257,7 +257,7 @@ const ProjectAnnualReportForm: React.FC<Props> = (props) => {
                   id={`form-${report.id}`}
                   validateOnMount={report.changeStatus === "staged"}
                   idPrefix={`form-${report.id}`}
-                  ref={(el) => (formRefs.current[report.id] = el)}
+                  ref={(el) => el && (formRefs.current[report.id] = el)}
```
[the remark this answers is not in the document]
Comment (pbastia) [reply app/components/Form/ProjectAnnualReportForm.tsx:260]: ...probably not in this PR though ;)

File: app/package.json [hunk 2]
```diff
@@ -103,33 +103,33 @@
...
     "openid-client": "^5.0.2",
     "path-to-regexp": "^6.2.0",
     "pg": "^8.7.1",
     "postgraphile": "^4.12.4",
     "postgraphile-log-consola": "^1.0.1",
     "postgraphile-plugin-connection-filter": "^2.2.2",
-    "react": "^17.0.2",
-    "react-datepicker": "^4.7.0",
-    "react-dom": "^17.0.2",
+    "react": "^18.2.0",
```
[the remark this answers is not in the document]
Comment (tmastrom) [reply app/package.json:123]: Yes, React minimum version was bumped to 18.2.0 for NextJS 13

File: app/cypress/support/commands.js [hunk 3]
```diff
@@ -571,3 +571,15 @@ Cypress.Commands.add("setDateInPicker", (ariaLabel, date, reportNumber = 0) => {
...
 });
+
+// This is a workaround for CI tests failing due to hydration errors but they are not failing locally
+Cypress.on("uncaught:exception", (err) => {
+  // we check if the error is
+  if (
+    err.message.includes("Minified React error #425;") ||
+    err.message.includes("Minified React error #418;") ||
+    err.message.includes("Minified React error #422;")
+  ) {
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/cypress/support/commands.js]: I was not too fond of this either but it was a workaround for passing cypress tests only on CI (local env was happy). Now that I've removed it from the code, e2e tests seem to be satisfied so I don't know what is going on here :) could be something related to bumping the nodejs version?!

File: app/components/Form/ProjectAnnualReportForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/components/Form/ProjectAnnualReportForm.tsx:260]: Tech debt card: https://app.zenhub.com/workspaces/climate-action-secretariat-60ca4121764d710011481ca2/issues/gh/bcgov/cas-cif/1551

File: app/tests/unit/components/Table/SortableHeader.test.tsx [hunk 4]
```diff
@@ -1,9 +1,9 @@
-import { useRouter } from "next/router";
 import { mocked } from "jest-mock";
 import React from "react";
 import SortableHeader from "components/Table/SortableHeader";
 import { render, screen } from "@testing-library/react";
-jest.mock("next/router");
+
+const useRouter = jest.spyOn(require("next/router"), "useRouter");
```
Comment (pbastia) [inline app/tests/unit/components/Table/SortableHeader.test.tsx:6]: I'm curious what error you were getting that was requiring this syntax, we prefer the `jest.mock` variant where we can

Comment (Sepehr-Sobhani) [reply app/tests/unit/components/Table/SortableHeader.test.tsx:6]: Sorry for not being descriptive in the commit message.
I was getting this error:
Screenshot 2023-03-16 at 12 51 15 PM
and the workaround was here: https://github.com/vercel/next.js/discussions/13891

Comment (Sepehr-Sobhani) [reply app/tests/unit/components/Table/SortableHeader.test.tsx]: I've found [this](https://github.com/vercel/next.js/issues/7479#issuecomment-533657701) which helps keep the `jest.mock`.

Comment (pbastia) [reply app/tests/unit/components/Table/SortableHeader.test.tsx]: Makes sense! Either way it works.

**Review by pbastia — APPROVED** [review]

**Pull request merged**